### ISSUE TITLE
Unify with swagger-js converter.

### ISSUE THREAD

**Issue by IvanGoncharov** [apigee-127/swagger-converter#26]
[swagger-js](https://github.com/swagger-api/swagger-js) project has it own [converter](https://github.com/swagger-api/swagger-js/blob/master/lib/spec-converter.js).
It part of private API and doesn't exposed to library users.
`swagger-js` used by `swagger-ui` this mean that converter heavily tested and more mature.

I compared both converters and found only one difference `swagger-js` download `apiDeclarations` on its  own. That also lead to async `convert` function.

@mohsen1 Could this two converters be merged into one?

**Comment by mohsen1** [comment]
The convertor in SwaggerJS does work with version 1.0 specs. It was written from scratch for that reason. We made this convertor prior to that for our own need and open sourced it. The file loading parts are not part of this repo because SwaggerTools is doing the heavy lifting for loading files. Turns our loading files in browser and server is it's own complex task so [there is a repo for that](https://github.com/whitlockjc/path-loader) which swagger tools is using.

**Comment by IvanGoncharov** [comment]
> The file loading parts are not part of this repo because SwaggerTools is doing the heavy lifting for loading files.

My idea is to take `swagger-js` converter and make it's API look like `swagger-converter` `convert` function. So made it sync and separate from file loader.

> The convertor in SwaggerJS does work with version 1.0 specs. It was written from scratch for that reason.

In reality people doesn't really distinguish between `1.0`, `1.1` and `1.2`.
Right now I'm trying to convert 140+ Swagger documents explicitly marked as `1.2` and it's total mess.
AFAIK, developers think if `swagger-ui` shows it correctly that mean swagger document is valid enough.

With current version of `swagger-converter` I managed to convert and validate(using `swagger-tools`) only 7 out of 140+ specs.

> We made this convertor prior to that for our own need and open sourced it.

I really like your library and it's already have community with developers open issues and RPs.
That is why I have idea to extend `swagger-convert` library and change `swagger-js` to use it.

API should stay the same and additional support for 1.0 and 1.1 should be backward-compatible.
I can do unification on per function basis by submitting PR.
And also adapt test-cases from `swagger-js`.

What do you think about such plan?

**Comment by mohsen1** [comment]
Keeping the API the same (sync and without worrying about loading) and adding other 1.x stuff support to this project is a great idea! I'm guessing @fehguy and @webron wont mind if you drop in a better replacement of existing convertor in swagger-ui/js. 

It's a good amount of work. I would say coordinate with @fehguy and make sure they are willing to accept a PR like that before starting the work. 

By the way I would like to learn more about the specs you couldn't convert using this project. What was missing?

**Comment by IvanGoncharov** [comment]
> By the way I would like to learn more about the specs you couldn't convert using this project. What was missing?

In general I found it really inefficient to fix issues based on output.
For example #27, was already fixed in `swagger-js`, see [here](https://github.com/swagger-api/swagger-js/blob/master/lib/spec-converter.js#L470).
Next issue was `int` type, it also already implemented [here](https://github.com/swagger-api/swagger-js/blob/master/lib/spec-converter.js#L413).
Tomorrow I will make an experiment, I will try to convert same specs using `swagger-js` converter.
I will put results here.

**Comment by IvanGoncharov** [comment]
By the way. I have very interesting statistics, in first bath I accumulate:

Version |Percent | Number
-----------|-----------|-------------
1.0|4.31%|17
1.1|23.09%|91
1.2|60.65%|239
2.0|12.69%|50

I didn't specifically target any particular version so results should represent real situation.
That mean that only ~13% is `2.0` specs, this make a real demand for reliable converter.
If we want to switch Swagger ecosystem to use `2.0` specification.

**Comment by IvanGoncharov** [comment]
Closing issue since I don't see too much 1.x Swagger specs anymore 🎉
